[filebeat] Add a Kafka input

faec opened the pull request:
Add a Kafka input to Filebeat (#7641).

urso commented on filebeat/input/kafka/input.go:
```diff
@@ -0,0 +1,211 @@
+	for msg := range claim.Messages() {
+		event := h.createEvent(sess, claim, msg)
+		h.input.outlet.OnEvent(event)
+		sess.MarkMessage(msg, "")
```
we really need to pass `util.Data`? Need to check if we can ignore the outleter and util.Data in new inputs.

Marking a message here does not give us end to end ACK. libbeat has support for end to end ACK, but filebeat architecture somehow conceals this. This is somewhat ok for the file input, but no other inputs. The way it is written here we are subject to data loss if events get not ACKed before filebeat shuts down.

kvch commented on filebeat/input/kafka/input.go:
```diff
@@ -0,0 +1,211 @@
+	return input, nil
+}
+
+func (p *Input) newConsumerGroup() (sarama.ConsumerGroup, error) {
```
I assume it's because previously inputs were named prospectors, and in other inputs receiver parameters were not renamed during migration.

faec replied:
Good question, I think I will change them to `input` :-)

urso commented:
For adding end-to-end ACK support you will need this: https://github.com/elastic/beats/pull/12997

end-to-end ACK becomes possible via:

```
    ...
    out, err := outlet.ConnectWith(cfg, beat.ClientConfig{
		Processing: beat.ProcessingConfig{
			DynamicFields: context.DynamicFields,
		},
        ACKEvents: func(privates []interface{}) {
            for _, priv := range []privates {
                if cm, ok := priv.(*sarama.ConsumerMessage); ok {
                    sess.MarkMessage(cm, "")
                }
            }
        },
	})
    ...

    var msg *sarama.ConsumerMessage
    ...
    out.OnEvent(beat.Event{
        Timestamp: ...
        Fields: ...
        Meta: ...
        Private: msg
    })
```

The ACKEvents callback will be called with the 'Private' field of all events recently ACKed. The pipeline ensures that ACKs are returned in the same order as you published your events.

Note: I have the sessions available in the callback. This is motivated by the fact that a pipeline.Client is not necessarily multithreading safe (we just made it safe because of the logs input in filebeat).

faec commented on filebeat/input/kafka/input.go:
```diff
@@ -0,0 +1,211 @@
+	consumerGroup, err :=
+		sarama.NewConsumerGroup(p.config.Hosts, p.config.GroupID, nil)
+	return consumerGroup, err
+}
```
oh yes, this isn't called anymore and was from an earlier iteration :-)

faec commented on filebeat/input/kafka/input.go:
```diff
@@ -0,0 +1,211 @@
+		go func() {
+			for err := range p.consumerGroup.Errors() {
+				p.log.Errorw("Error reading from kafka", "error", err)
+			}
```
From my best understanding of sarama:
- Errors here do not necessarily indicate dropped events, though I think they can depending on retry settings. sarama doesn't deliver any errors to the actual listener (in ConsumeClaim), only successfully received events, while communication / cluster errors are reported on this channel.
- consumerGroup.Consume only produces errors related to joining the group in the kafka cluster, not to the actual consumption process

faec commented on filebeat/input/kafka/input.go:
```diff
@@ -0,0 +1,211 @@
+				err := p.consumerGroup.Consume(p.kafkaContext, p.config.Topics, handler)
+				if err != nil {
+					p.log.Errorw("Kafka consume error", "error", err)
+				}
```
I don't believe so. Errors calling `Consume` should mean failure to join the consumer group to begin with, so there's no state to recover and all we can do is retry.

urso replied:
Why can clients fail to join a consumer group? Permissions, network failures, topic does not exist (yet)?

faec replied:
The errors that can be returned here are:

```
ErrClosedConsumerGroup
ErrInvalidTopic
ErrTopicAuthorizationFailed
ErrUnknownTopicOrPartition
ErrLeaderNotAvailable
ErrOutOfBrokers
ErrConsumerCoordinatorNotAvailable
```

(as far as I can tell there's no centralized list of possible errors in either the docs or the code, so I might have missed one, but this is everything I know of)

urso replied:
This is interesting, and how this actually interacts with the filebeat runner.
Currently the runner executes `Run` once. If Run returns and `--once` CLI flag is set by the user, then filebeat will return. But `--once` is meant to collect and stop once it reached EOF. Here `--once` forces a shutdown if you encountered an error in `Consume`.
If the user did not configure `--once`, then filebeat will wait for `scan_frequency` seconds (yep, your kafka input has a `scan_frequency` setting), and call `Run` again. Some inputs start a another go-routine. This is why they guard themselves from extra calls to Run. This doesn't play well with these inputs, cause filebeat and the runner think that the input has finished if `--once` is set, but this is not true...

That is, only by chance filebeat will retry if `Consume` fails here. Yet I'd like to remove this behavior from the runner (which behavior is specific/correct for the logs input only), or have an alternative runner.

If an input is statically configured, but returns/fails, then filebeat (normally) should shut down. This is also the case if your constructor fails. e.g. NewConsumerGroup might fail, and therefore force a filebeat shutdown if no other input is configured. With filebeat being a service, it should retry (with backoff) if an error is recoverable/healable.
For the errors given I'd say all can be resolved while filebeat is running, as all seem to be triggered by the outside. e.g. `ErrLeaderNotAvailable`, `ErrConsumerCoordinatorNotAvailable`, `ErrConsumerCoordinatorNotAvailable`, or `ErrConsumerCoordinatorNotAvailable` might indicate network errors or kafka cluster being unstable. `ErrInvalidTopic` might eventually resolved once a producer comes online. `ErrUnknownTopicOrPartition` is not helpful. While a topic might be unknown, it might get generated. Having the wrong partition might be more severe, but this should only occur if someone removed partitions.

Question: when can NewConsumerGroup error? I guess if the config is wrong, but also if connection attempt fails? In the former case we should return, but the later one would be 'recoverable'.

faec replied:
NewConsumerGroup only errors on configuration as far as I know, since it doesn't actually establish the connection it only creates the interface for it. The consumer group isn't actually joined until calling Consume.

In any case, alright, I will guard against multiple calls for now :-)

faec replied:
I have to amend this after more testing, the initial call to `NewConsumerGroup` in `NewInput` does actively join the group and will return failure if it can't reach the brokers. In light of that, I think it is better to wait to call `NewConsumerGroup` until `input.Run`, so our retry loop includes the initial connection, otherwise an ephemeral network error on startup could require a restart of the beat... oops.

faec commented on filebeat/input/kafka/input.go:
```diff
@@ -0,0 +1,211 @@
+		})
+	}
+	return array
+}
```
I thought about this, but as far as I can tell duplicate header keys are valid (if not in the spec then at least in the implementation -- when I post events with duplicate headers, it passes them on unchanged to the consumer)

urso replied:
Checking the kafka implementation it is indeed just a list of headers, allowing for duplicates.
This is not optimal for querying/filtering events via kibana. A simple map[string]string would fit kibana UI better. I wonder if `map[string][]string` might be a better compromise.

urso commented on filebeat/input/kafka/kafka_integration_test.go:
```diff
@@ -0,0 +1,251 @@
+}
+
+func strDefault(a, defaults string) string {
+	if len(a) == 0 {
```
nit; `if a == "" {` is equivalent.

urso commented:
fyi, this one adds some support for context.Context to channel.Outlet: https://github.com/elastic/beats/pull/13031

faec commented on filebeat/input/kafka/input.go:
```diff
@@ -0,0 +1,211 @@
+
+// Wait shuts down the Input by cancelling the internal context.
+func (p *Input) Wait() {
+	p.Stop()
```
Yes, this wasn't right. I added a wait for the kafka consumer group to close now, which is closer to correct, but I think this part won't be finalized until I add your end-to-end support.

`p.started` was leftover from code copied from other inputs... it doesn't serve a purpose here anymore except as a non-thread-safe `Once` on `Run()`, so I've removed it. (I checked and e.g. the log input has no such safeguard, though if that behavior is still desired, I can add it back with a real `Once` :-) )

urso replied:
Unfortunatelly you have to guard your Run method from multiple executions. Looks like this is the purpose of `p.start`. The runner might execute the Run method every now and so often (default every 10s). But only the logs input makes use of it.
It's kinda stupid. I will attempt to fix this next, then we can remove the guard.

faec commented on filebeat/input/kafka/config.go:
```diff
@@ -0,0 +1,147 @@
+	TLS           *tlscommon.Config `config:"ssl"`
+	Username      string            `config:"username"`
+	Password      string            `config:"password"`
+}
```
👍 to most of those, though our vendored kafka doesn't yet support `IsolationLevel` (I can add upgrading it to my task list if we'd like that to happen soon)

urso replied:
+1 on upgrading

faec replied:
The upgrade is merged, I just added IsolationLevel to this PR as well

urso commented:
fyi; #12997 has been merged. The next time you rebase you have to update the inputs constructor.

faec commented:
I've added end-to-end support. It took some synchronization work to make the session available to the ACKer, let me know if there's a cleaner way to do it with your refactor but this should guard against conflicts between ACKs and session state changes. I believe the main outstanding question left is how we should represent headers in the index.

urso commented on filebeat/docs/inputs/input-kafka.asciidoc:
```diff
@@ -0,0 +1,107 @@
+
+To configure this input, specify a list of <<hosts,`hosts`>> to use for this
+cluster, a list of <<topics,`topics`>> to track, and a <<groupid,`group_id`>>
+to connect with.
```
We should note that the brokers mentioned are only used for bootstrapping the actual connections. One doesn't need to enter all hosts of a cluster here.

urso commented on filebeat/docs/inputs/input-kafka.asciidoc:
```diff
@@ -0,0 +1,107 @@
+[[hosts]]
+===== `hosts`
+
+A list of Kafka hosts (brokers) for this cluster.
```
'bootstrapping hosts'. Connection is established based on cluster metadata

urso commented on filebeat/input/kafka/config.go:
```diff
@@ -0,0 +1,220 @@
+	TLS           *tlscommon.Config `config:"ssl"`
+	Username      string            `config:"username"`
+	Password      string            `config:"password"`
+}
```
nit: I noticed `Fetch` and `Rebalance` are already set in the default config. In this case there is no reason to use pointers. Code normally uses pointer to check if a config (section) is fully optional, and we don't even care about defaults if it not given at all.

urso commented on filebeat/input/kafka/input.go:
```diff
@@ -0,0 +1,285 @@
+	sessionState  *kafkaSessionState
+	log           *logp.Logger
+	runOnce       sync.Once
+}
```
do we need all fields? E.g. I don't see rawConfig being used.

faec replied:
No, it was used in an earlier iteration but can be removed now

urso commented on filebeat/input/kafka/input.go:
```diff
@@ -0,0 +1,285 @@
+
+// Stop shuts down the Input by cancelling the internal context.
+func (input *kafkaInput) Stop() {
+	close(input.context.Done)
```
I think `filebeat/input.Runner` already closes the channel on stop. If so, we have a chance of creating a panic here.

I think long-term 'Stop' becomes obsolete, if we make more use of context (or similar) to signal shutdown to run loops.

faec replied:
It took a while to convince myself that they're the same channel and that the program flow always closes it, but I think you're right. I agree the `Stop` mechanism is kind of begging for a refactor right now and would probably be better off clarified and / or removed (since it's hard to follow who's responsible for closing which resource).

For now I've corrected this to closing `input.outlet`, and added a comment clarifying why explicitly closing the kafka consumer group is redundant here.

urso commented on filebeat/input/kafka/input.go:
```diff
@@ -0,0 +1,285 @@
+// as a utility in a shared part of the code, but right now it's a special
+// case for sarama which requires a context.Context, so it's private until
+// there's at least one other use case.
+type channelCtx <-chan struct{}
```
Please create a follow up issue to clear this in the future. I think the S3 input is adding the same type.

@ph did add some close signaling pattern to go-concert: https://github.com/elastic/go-concert/pull/7
It's more focused on signaling shutdown by building a tree of dependencies + calling functions during shutdown if required (e.g. to close sockets). I think it would make sense for his type to be castable to context.Context.
Consider to add this pattern for reuse to go-concert. Historically we have a many done channels all over the place. Would be nice if we can pass them down into IO libs as context.Context from other places as well.

urso commented on filebeat/input/kafka/input.go:
```diff
@@ -0,0 +1,285 @@
+				default:
+				}
+			}
+		}()
```
I think you can simplify the loop to:

```
context := doneChannelContext(input.context.Done)
handler := groupHandler{input: input}
for context.Err() == nil {
  err := input.consumerGroup.Consume(context, ...)
  if err != nil {
    input.log.Errorw(...)
  }
}
```

Question: do we need some backoff between errors?

faec replied:
Yes, this is much prettier. And yes, I think I will add backoff here along with the other pending run loop fixes.

urso commented on filebeat/input/kafka/input.go:
```diff
@@ -0,0 +1,285 @@
+	})
+	if err != nil {
+		return nil, err
+	}
```
The ACKEvents handler can be called for as long as the underlying `beat.Client` is still connected. I guess there is a high chance that the sarama user group closes before the `beat.Client`, giving us ACKs for a session that has been closed already. Looks like this is the also a reason you have to check for `nil` session for every event.

I wonder if the session is bound to a handler. In this case 'Private' could be a tuple of handler+event. All logic about tracking the liveness of a session + ACKing messages would be local to the handler then, reducing the need for some 'global' sharing across instances of different types.

e.g.

```
type eventExtra struct {
  handler *groupHandler
  message *sarame.ConsumerMessage
}

...

ACKHandler: func(privates []interface{}) {
  for _, priv := range privates {
    if extra, ok := priv.(eventExtra); ok {
      extra.handler.ACK(extra.message)
    }
  }
}
```

WDYT?

The ClientConfig itself supports a 'WaitShutdown' setting, to postpone shutdown for a little. If configured `Close` will wait until all events have been ACKed or after until the configured timeout. It would be nice to have the same functionality via a custom delaying context and expose this functionality via configs to users. This would reduce the chance of duplicates incurred by shutdown. We can add this functionality in a later PR.

urso replied:
The groupHandler seems to be 'our' side of the session. If we protect Start, Cleanup, and ACK methods using a mutex, then we would have the session sync, handling + acking in one type only.

faec replied:
One thing I wasn't confident of on the first pass, but I realize you've since confirmed, is that ACKs will be sent sequentially, so that in particular only one ACK can be executed at a time (so the wait group is superfluous and the whole system can operate on a single mutex). Combined with the mistake mentioned in another comment about network errors returned from `NewConsumerGroup`, the run loop / ack logic needs revision. I will get you a simpler and more correct version soon.

urso commented on filebeat/input/kafka/input.go:
```diff
@@ -0,0 +1,285 @@
+	for msg := range claim.Messages() {
+		event := h.createEvent(sess, claim, msg)
+		h.input.outlet.OnEvent(event)
+		sess.MarkMessage(msg, "")
```
I'm confused, do we still need to call 'MarkMessage' here?

faec replied:
nope, this just didn't get updated >_< thanks!

vjsamuel commented:
@urso @ph, @faec  can we have this PR modeled similar to how tcp is modeled? can we have a Kakfa inputsource which can take a callback that can take a Kafka message, process it and add them to the outlet?

that allows folks to take arbitrary messages from Kafka and convert them to events.

urso commented:
@vjsamuel In general I'm in favor of separating the 'source' from the 'processing', as we do it for the TCP and UDP input.

What would be the use-case you have in mind? With kafka messages have clear boundaries. Do you plan to parse and split the message into multiple events? Wouldn't it make sense to provide reusable processors actual users can configure?

vjsamuel commented on filebeat/input/kafka/input.go:
```diff
@@ -0,0 +1,281 @@
+		Timestamp: time.Now(),
+	}
+	eventFields := common.MapStr{
+		"message": string(message.Value),
```
@urso , this is why i propose to make it an input source instead of just having a kafka input. `message.Value` being a string is a generalization. in many cases Kafka messages can be written in as Protobuf/Avro messages that would need to be deserialized before the event is written out. a user must be able to provide what Decoder to use on Sarama and how to create an event. kindly consider this use case.

cc: @ph

faec commented:
I've revised the core run loop to be more robust, fixed end-to-end ACK per @urso's suggestions about `groupHandler`, and fixed a shutdown bug that I came across in the process. The integration test now waits for the input to shutdown and confirms it takes <30sec.

I also looked at the header issue in a live index, and currently it's definitely not what we want:

```
curl "http://localhost:9200/filebeat-8.0.0-2019.08.01-000001/_search?pretty"
...
            "headers" : [
              {
                "key" : [
                  107,
                  101,
                  121,
                  115,
                  32,
                  97,
                  110,
...
```

I initially assumed a data blob was the safest since that's what's in the spec, but given this outcome, absent better suggestions, I'll make them strings and leave it to the receiver to re-extract a data blob if that's truly their intent.

urso commented on filebeat/input/kafka/kafka_integration_test.go:
```diff
@@ -0,0 +1,265 @@
+	}
+
+	// Setup the input config
+	config, _ := common.NewConfigFrom(common.MapStr{
```
Use `MustNewConfigFrom`. This would panic, potentially pointing out programmer error when setting up the test.

faec commented:
Update on the headers issue, following several tests and an offline discussion: the suggestion above about writing them to `map[string][]string` doesn't allow for the kind of searching we'd like. The current container, `[]struct{ key, value: string }`, can work in theory as long as it's indexed as a `nested` type. I've updated `filebeat/_meta/fields.common.yml` with something near the correct config, but for some reason the properties inside `headers` (`key`, `value`) do not appear in the index template.

I can create index templates for fields like this manually, but when the template is built by filebeat the subfields are omitted. None of our beats / modules / etc currently use a `nested` type with defined subfields, so it's not clear to me yet whether the definition in `fields.common.yml` is somehow incorrect, or if the template generator doesn't handle subfields of a `nested` field.

faec commented on filebeat/input/kafka/input.go:
```diff
@@ -0,0 +1,293 @@
+				select {
+				case <-input.context.Done:
+					return
+				case <-time.After(input.config.ConnectBackoff):
```
The `backoff` package works only based on the `Wait` function, not a channel, so I had to wrap it in an auxiliary channel to handle shutdown signals, let me know if you see a prettier approach.

urso replied:
Internally the implementation uses a channel. We could introduce a method like `C() <- chan time.Time`, which is created by a timer. If the current timer is not nil, it should be stopped and reset when calling `C`.

faec commented on filebeat/input/kafka/input.go:
```diff
@@ -0,0 +1,293 @@
+	inputContext input.Context,
+) (input.Input, error) {
+
+	out, err := connector.ConnectWith(cfg, beat.ClientConfig{
```
Added the CloseRef parameter (which also let me remove the extra Once for shutdown). I'm not sure how to use `WaitClose` effectively though, since I don't see any code that uses it currently. Are you suggesting I make it configurable as part of the input via `kafkaInputConfig`?

urso replied:
for WaitClose you only have to configure a duration. When the outlet is closed, it blocks after all events have been ACKed, or until the configured duration passed. During this period ACKs are still received, but no events can be published anymore.

> Are you suggesting I make it configurable as part of the input via kafkaInputConfig?

Yes.

urso commented on filebeat/input/kafka/input.go:
```diff
@@ -0,0 +1,322 @@
+					backoff.Wait()
+				}()
+
+				select {
```
you already initialized the backoff instance with the same done channel. Calling `Wait` should give you exactly the same behavior.

urso commented on filebeat/input/kafka/input.go:
```diff
@@ -0,0 +1,315 @@
+				select {
+				case <-input.context.Done:
+					return
+				}
```
this is missing the `default:` clause, otherwise it blocks forever.
You already created a `context.Context` for the consumer group. If you pass it to the run loop (or generate it again, it's just a type cast), then the for loop becomes:

```
for context.Err() == nil {
   err := input.runConsumerGroup()
   backoff.Wait()
}
```

Where do you call `backoff.Reset()`, so to reset the wait duration to the initial wait time?

faec replied:
I didn't call reset in this code -- this was somewhat jumbled because that wait was applying to any fatal error throughout the run loop, so there was no natural place to reset it. I've redistributed the helper code again -- the connection backoff now applies exclusively to the creation / initial connection of the consumer group, and is reset when it succeeds. Errors during the consumption process are subject to sarama's own retry / backoff settings, so when any entire session is lost we just restart the initial connection process right away. Does this make sense?

urso replied:
Sounds good.

urso commented:
@dedemorton Can you have a look at the docs?

jsoriano commented on filebeat/_meta/fields.common.yml:
```diff
@@ -150,3 +150,37 @@
+        - name: partition
+          type: long
+          description: >
+            Kafka partition number
```
@faec @urso I wonder if we should use the same fields metricbeat uses for this, in case we want to correlate and/or monitor the kafka queues used. Metricbeat uses `kafka.topic.name` for the topic and `kafka.partition.id` for the partition.

And I wonder if message specific fields (`offset`, `key`, `headers`) should be under an specific namespace as `kafka.message`, at least to differentiate the offset from other offsets.
